Add Slang shaders to additional samples

SaschaWillems opened the pull request:
## Description

**Important note:** If you compile shaders locally, you have to use a very recent version of the slang compiler. The one shipped with the SDK is outdated and broken. Make sure to grab the latest one from https://github.com/shader-slang/slang/releases 

This adds Slang shaders to additional samples:

1. buffer device address
2. calibrated timestamps
3. color write enable
4. compute nbody
5. compute nbody (hpp)
6. conservative rasterization
7. conditional rendering
8. debug utils
9. descriptor buffer basic
10. descriptor indexing
11. dynamic blending
12. dynamic line rasterization
13. dynamic multisample rasterization
14. dynamic primitive clipping
15. dynamic uniform buffers
16. dynamic uniform buffers (hpp)
17. extended dynamic state2
18. fragment shader barycentric
19. fragment shading rate
20. graphics pipeline library
21. gshader to mshader
22. hdr
23. hdr (hpp)
24. host image copy
25. instancing
26. instancing (hpp)
27. logic op dynamic state
28. memory budget
29. mesh shading
30. mesh shading (hpp)
31. open cl interop
32. open gl interop
33. patch control points
34. push descriptors
35. push descriptors (hpp)
36. portability
37. ray queries
38. ray tracing basics
39. ray tracing extended
40. ray tracing position fetch
41. separte image sampler
42. separte image sampler (hpp)
43. shader debug printf
44. sparse image
45. synchronization 2
46. terrain tessellation
47. terrain tessellation (hpp)
48. texture compression basisu
49. texture loading
50. texture loading (hpp)
51. texture mipmap generation
52. texture mipmap generation (hpp)
53. timeline semaphores
54. timestamp queries
55. timestamp queries (hpp)
56. vertex dynamic state

Also enables SPIR-V 1.4 at framework level when using Slang.

Additional notes:
* Adds a new option `VKB_SKIP_SLANG_SHADER_COMPILATION` to skip compilation of Slang shaders (you can still use Slang shaders via the precompiled SPIR-V files)
* Disables Slang offline compilation in CI for MacOS/iOS due to broken Slang compiler in SDK (this step is unnecessary anyway)
* Some samples may look slightly different than their GLSL counterparts, that's not (easily) fixable

Refs #1385

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [x] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

SaschaWillems committed:
Enabled SPIR-V 1.4 extension and use VK 1.1 if required

This is needed to get proper support for SPIR-V compiled from Slang shaders

SaschaWillems committed:
Add slang shaders for

- instancing
- conditional rendering
- conservative rasterization

SaschaWillems committed:
Add slang shaders for

- dynamic uniform buffers
- descriptor buffer basic
- fragment shading rate

SaschaWillems committed:
Add slang shaders for

- hdr (+hpp)
- graphics pipeline library
- mesh shading (+hpp)

SaschaWillems committed:
Fix copyright

SaschaWillems committed:
Add slang shaders for

- texture loading (+hpp)
- push descriptors (+hpp
- open cl interop

SaschaWillems committed:
Fix copyright

SaschaWillems committed:
Add slang shaders for

- separte image sampler (+hpp)
- texture mipmap generation (+hpp)
- shader debug printf
- texture compression basisu

SaschaWillems committed:
Add slang shaders for

- compute nbody (+hpp)

SaschaWillems committed:
Add slang shaders for

- dynamic primitive clipping
- logic op dynamic state
- sparse image

SaschaWillems committed:
Add slang shaders for

- color write enable
- debug utils
- dynamic line rasterization

SaschaWillems committed:
Add slang shaders for

- ray queries
- ray tracing basics
- ray tracing extended
- ray tracing position fetch

SaschaWillems committed:
Update reeadme for shaders

SaschaWillems committed:
Add slang shaders for

- terrain tessellation
- patch control points
- synchronization 2
- vertex dynamic state

SaschaWillems committed:
Add slang shaders to hpp variant

SRSaunders commented:
It looks like two issues are occurring here with Apple CI:

1. PR #1405 will fix CI issues with macOS (and iOS) when using runners with the latest Sequoia SDKs, as is here.
2. The iOS CI is also failing because it has access to **slang** from the Vulkan SDK, and generates an error when compiling the sparse image shader (note: sparse images are not yet supported on Apple):

_/Users/runner/work/Vulkan-Samples/Vulkan-Samples/shaders/sparse_image/slang/sparse.frag.slang(45): error 36107: entrypoint 'main' does not support compilation target 'spirv' with stage 'fragment'
float4 main(VSOutput input)
       ^~~~
/Users/runner/work/Vulkan-Samples/Vulkan-Samples/shaders/sparse_image/slang/sparse.frag.slang(56): note: see using of 'CheckAccessFullyMapped'
    } while (!CheckAccessFullyMapped(residencyCode));
** BUILD FAILED **
              ^~~~~~~~~~~~~~~~~~~~~~_

Also note that macOS CI does not access the VulkanSDK and will not try to compile any slang shaders, but would likely fail on the same sparse image shader if it did.

SaschaWillems commented:
It's a draft for a reason.

SRSaunders commented:
Understood - just trying to help out re issue #1405 since this has nothing to do with your PR, only the Apple CI.

SaschaWillems commented:
You're welcome. We discussed that on one of our calls, and having each CI step rebuild the SPIR-V files that are already part of the repo doesn't make sense anyway. Esp. as the SDK tends to include broken/outdated compilers.

SRSaunders commented:
> having each CI step rebuild the SPIR-V files that are already part of the repo doesn't make sense anyway

I agree.

However, the scope of the problem may be limited to iOS.  At this point iOS seems to be the only CI target that requires and accesses the Vulkan SDK for _static linkage_ to the MoltenVK framework.  I believe all other CI targets depend on Vulkan-Headers only without need for the SDK (or compilers), assuming _dynamic linkage_ to the Vulkan runtime library.

SaschaWillems commented:
Should put that in a separate issue.

I'd like the comments section of this PR related to the actual PR if possible.

SaschaWillems committed:
Add slang shaders for

- descriptor indexing

SaschaWillems committed:
Add slang shaders for

- dyamic blending

SaschaWillems committed:
Add slang shaders for

- time stamp queries (+hpp)
- host image copy

SaschaWillems committed:
Add slang shaders for

- fragment shader barycentric
- memory budget

SaschaWillems committed:
Add slang shaders for

- calibrated timestamps
- dynamic multisample rasterization

SaschaWillems committed:
Add slang shaders for

- open gl interop

SaschaWillems committed:
Add slang shaders for

- portability

SaschaWillems committed:
Fix copyright

SaschaWillems committed:
Skip Slang offline compilation on MacOS/iOS when run through CI

Slang compiler in current SDK is broken
Not necessary anway as we ship offline compiled SPIR-V anyway, so unnecessary waste of CI time

SaschaWillems committed:
Fix copyright

SRSaunders commented:
@SaschaWillems I have tested this PR on macOS and iOS (collectively "Apple") and it runs, but have learned the following:

1. Vulkan SDK 1.4.321.0 (the most recent) is required for successful slangc shader compilation (slangc version 2025.11-12-gc5295eae2).  Earlier SDKs and slangc compilers don't work, at least on Apple.  You are correct in your assessment about broken compilers in earlier SDKs.
2. Certain samples emit the following validation layer error on Apple but continue to operate.   I am not sure if this is an Apple-only issue, or if this may be a validation layer false positive:

_[error] 115483881 - VUID-VkShaderModuleCreateInfo-pCode-08740: vkCreateShaderModule(): SPIR-V Capability DrawParameters was declared, but one of the following requirements is required (VkPhysicalDeviceVulkan11Features::shaderDrawParameters OR VK_KHR_shader_draw_parameters).
The Vulkan spec states: If pCode is a pointer to SPIR-V code, and pCode declares any of the capabilities listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://vulkan.lunarg.com/doc/view/1.4.321.0/mac/antora/spec/latest/chapters/shaders.html#VUID-VkShaderModuleCreateInfo-pCode-08740)_

3. The above error is strange since since `VkPhysicalDeviceVulkan11Features::shaderDrawParameters` is true on Apple for Core 1.1.  However, if I add `VK_KHR_SHADER_DRAW_PARAMETERS_EXTENSION_NAME` to the enabled extensions for slang, then the validation error disappears.
4. Certain samples render incorrectly with slang (e.g. _descriptor_indexing_, _host_image_copy_), at least on Apple.

SaschaWillems commented:
Thanks for testing your feedback 👍🏻. I'm aware of both issues (validation and visual differences).

As for the SDK: Slang moves so fast, that I haven't seen a good version in any of the SDKs. Even the one in the latest SDK does have issues, that don't show yet for our samples, as the shaders converted up to now are mostly simplistic. 

Descriptor indexing is fixed. Other samples like host image copy also look wrong in HLSL, fixing them will be done outside of this PR at a later point.

SaschaWillems committed:
Add slang shaders for

- gshader_to_mshader

Still needs some lighting fixes

SaschaWillems committed:
Fix lighting for gshader_to_mshader

Calculate normal matrix on CPU

SaschaWillems committed:
Add slang shaders for

- extended dynamic state2

SaschaWillems committed:
Fix descriptor indexing not properly displaying right side of screen when using Slang

SaschaWillems committed:
Add offline compiled SPIR-V files for slang shaders

SaschaWillems committed:
Enalbe VK_KHR_Shader_Draw_Parameters when using Slang

SaschaWillems committed:
Add slang shaders for

- timeline semaphores

SaschaWillems committed:
Add slang shaders for

- buffer device address

SaschaWillems committed:
Merge branch 'main' into shaders_09_2025

SRSaunders commented:
> As for the SDK: Slang moves so fast, that I haven't seen a good version in any of the SDKs. Even the one in the latest SDK does have issues, that don't show yet for our samples, as the shaders converted up to now are mostly simplistic.

@SaschaWillems while CI now passes for macOS and iOS (due to skipping slanc compilation), interactive compilation of the slang shaders has started to fail since your _descriptor_indexing_ fix.  This is even with the most recent SDK (1.4.321). Unfortunately this breaks interactive builds on macOS and iOS.  As a short term solution I wonder if you should just disable slangc on Apple for all build types until things stabilize in the SDK,  i.e. remove the term `($ENV{CI} MATCHES true) AND` from **sample_helper.cmake**?

Alternatively, a better improvment might be to have a global cmake config switch that disables all interactive shader compilation (glsl, hlsl, slang) by default.  You have offline-compiled all shaders and this should be fine for most people.  A user could then enable the switch if they really want to recompile the shaders.  Note that on Apple, the build system currently recompiles all the shaders for all languages all the time - a waste for most people.

SaschaWillems commented:
Please see the issue description. The Slang compiler in the latest SDK is downright broken.

Disabling shader compilation via a switch, if desired, should be done outside of this PR. My job (with this) was to add Slang shaders.

We're currently discussing these issue on a larger scale inside the WG.

SRSaunders commented:
> The Slang compiler in the latest SDK is downright broken.

I get it.  However, if you merge this PR all Apple interactive builds will be broken.  Could you at least fix this by disabling slangc all the time on Apple?

SaschaWillems commented:
This won't be merged until we (inside the samples group) agree. This includes the apple platform discussion.

SaschaWillems commented:
Created #1409 to make sure your remarks won't get lost.

SRSaunders commented:
Thanks.  I too am in favour of a general solution that works across platforms and shading languages.

asuessenbach commented on shaders/buffer_device_address/slang/update_vbo.comp.slang:
```diff
@@ -0,0 +1,70 @@
+void main(uint3 globalInvocationID: SV_DispatchThreadID, uint3 workgroupID: SV_GroupID, uniform Registers registers)
+{
+    uint3 workgroupSize = WorkgroupSize();
+    uint3 workgroupCount = WorkgroupCount();
```
With VS2022, I get `error 30015: undefined identifier 'WorkgroupCount'.` here.

SaschaWillems replied:
Please update your Slang compiler (see issue description).

WorkgroupCount was only recently added, and e.g. the compiler from the SDK does not have that function.

gary-sweet commented:
I don't have a local slang compiler, so I should be running against the pre-built versions.

I am seeing some validation errors (and some other problems which I'll share afterwards). These may, of course, not be anything to do with the slang, but could be compilers problems or spirv-val problems. I thought I'd list them anyway.

**compute_nbody**
`[error] -1307510846 - VUID-StandaloneSpirv-None-10684: vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Invalid explicit layout decorations on type for operand '141[%141]'
  %sharedData = OpVariable %_ptr_Workgroup__arr_v4float_SHARED_DATA_SIZE Workgroup
The Vulkan spec states: All variables must have valid explicit layout decorations as described in Shader Interfaces (https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-StandaloneSpirv-None-10684)`

`[error] 1132206547 - VUID-VkPipelineShaderStageCreateInfo-pSpecializationInfo-06849: vkCreateComputePipelines(): pCreateInfos[0].stage After specialization was applied, VkShaderModule 0x500000000050 produces a spirv-val error (stage VK_SHADER_STAGE_COMPUTE_BIT):
[VUID-StandaloneSpirv-None-10684] Invalid explicit layout decorations on type for operand '141[%141]'
  %sharedData = OpVariable %_ptr_Workgroup__arr_v4float_SHARED_DATA_SIZE Workgroup
The Vulkan spec states: If a shader module identifier is not specified, the shader code used by the pipeline must be valid as described by the Khronos SPIR-V Specification after applying the specializations provided in pSpecializationInfo, if any, and then converting all specialization constants into fixed constants (https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-VkPipelineShaderStageCreateInfo-pSpecializationInfo-06849)`

**hello_triangle**
`[error] -1520283006 Validation Layer: Error: VUID-VkShaderModuleCreateInfo-pCode-08737: vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Invalid SPIR-V binary version 1.4 for target environment SPIR-V 1.3 (under Vulkan 1.1 semantics).
The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must adhere to the validation rules described by the Validation Rules within a Module section of the SPIR-V Environment appendix (https://docs.vulkan.org/spec/latest/chapters/shaders.html#VUID-VkShaderModuleCreateInfo-pCode-08737)`

**hpp_compute_nbody**
`[error] -1307510846 - VUID-StandaloneSpirv-None-10684: vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Invalid explicit layout decorations on type for operand '141[%141]'
  %sharedData = OpVariable %_ptr_Workgroup__arr_v4float_SHARED_DATA_SIZE Workgroup
The Vulkan spec states: All variables must have valid explicit layout decorations as described in Shader Interfaces (https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-StandaloneSpirv-None-10684)`

`[error] 1132206547 - VUID-VkPipelineShaderStageCreateInfo-pSpecializationInfo-06849: vkCreateComputePipelines(): pCreateInfos[0].stage After specialization was applied, VkShaderModule 0x500000000050 produces a spirv-val error (stage VK_SHADER_STAGE_COMPUTE_BIT):
[VUID-StandaloneSpirv-None-10684] Invalid explicit layout decorations on type for operand '141[%141]'
  %sharedData = OpVariable %_ptr_Workgroup__arr_v4float_SHARED_DATA_SIZE Workgroup
The Vulkan spec states: If a shader module identifier is not specified, the shader code used by the pipeline must be valid as described by the Khronos SPIR-V Specification after applying the specializations provided in pSpecializationInfo, if any, and then converting all specialization constants into fixed constants (https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-VkPipelineShaderStageCreateInfo-pSpecializationInfo-06849)`

**hpp_texture_mipmap_generation**
`[error] -1307510846 - VUID-StandaloneSpirv-None-10684: vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Invalid explicit layout decorations on type for operand '24[%24]'
  %samplers = OpVariable %_ptr_UniformConstant__arr_23_int_3 UniformConstant
The Vulkan spec states: All variables must have valid explicit layout decorations as described in Shader Interfaces (https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-StandaloneSpirv-None-10684)`

**color_write_enable**
`[error] 115483881 - VUID-VkShaderModuleCreateInfo-pCode-08740: vkCreateShaderModule(): SPIR-V Capability StorageImageReadWithoutFormat was declared, but one of the following requirements is required (VkPhysicalDeviceFeatures::shaderStorageImageReadWithoutFormat OR VK_VERSION_1_3 OR VK_KHR_format_feature_flags2).
The Vulkan spec states: If pCode is a pointer to SPIR-V code, and pCode declares any of the capabilities listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://docs.vulkan.org/spec/latest/chapters/shaders.html#VUID-VkShaderModuleCreateInfo-pCode-08740)`

**synchronization_2**
`[error] -1307510846 - VUID-StandaloneSpirv-None-10684: vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Invalid explicit layout decorations on type for operand '141[%141]'
  %sharedData = OpVariable %_ptr_Workgroup__arr_v4float_SHARED_DATA_SIZE Workgroup
The Vulkan spec states: All variables must have valid explicit layout decorations as described in Shader Interfaces (https://docs.vulkan.org/spec/latest/appendices/spirvenv.html#VUID-StandaloneSpirv-None-10684)`

`[error] 1132206547 - VUID-VkPipelineShaderStageCreateInfo-pSpecializationInfo-06849: vkCreateComputePipelines(): pCreateInfos[0].stage After specialization was applied, VkShaderModule 0xf2f0000000f2f produces a spirv-val error (stage VK_SHADER_STAGE_COMPUTE_BIT):
[VUID-StandaloneSpirv-None-10684] Invalid explicit layout decorations on type for operand '141[%141]'
  %sharedData = OpVariable %_ptr_Workgroup__arr_v4float_SHARED_DATA_SIZE Workgroup
The Vulkan spec states: If a shader module identifier is not specified, the shader code used by the pipeline must be valid as described by the Khronos SPIR-V Specification after applying the specializations provided in pSpecializationInfo, if any, and then converting all specialization constants into fixed constants (https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-VkPipelineShaderStageCreateInfo-pSpecializationInfo-06849)`

gary-sweet commented on shaders/hdr/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,127 @@
+			fresnel *= (1.0 - F0);
+			fresnel += F0;
+
+			float spec = (fresnel * geoAtt) / (NdotV * NdotL * 3.14);
```
This seems to be missing the clamping that the glsl version has:
```
// Note: clamp to zero to mitigate any divide by zero
float spec = max((fresnel * geoAtt) / (NdotV * NdotL * 3.14), 0.0);
```

gary-sweet replied:
Looks like the hlsl version is wrong too to be fair.

SaschaWillems replied:
Fixed for both HLSL and Slang

gary-sweet commented on shaders/conditional_rendering/slang/model.frag.slang:
```diff
@@ -0,0 +1,38 @@
+	float3 L = normalize(input.LightVec);
+	float3 V = normalize(input.ViewVec);
+	float3 R = reflect(-L, N);
+	float3 ambient = float3(0.1);
```
The ambient was 0.25 in glsl and hlsl, so this model looks darker

SaschaWillems replied:
Fixed

SaschaWillems committed:
Recompile Slang shader with updated compiler version

SaschaWillems commented:
Thanks for testing 👍🏻

Most of the validation errors were caused by the Slang version that I used generating invalid SPIR-V. I have recompiled the shaders with a newer versions and most samples should now be free of validation errors.

For some like hello triangle and color write enable, the samples need special handling when using Slang shaders due to a slightly different shading environment.

Will try to fix these next.

SaschaWillems committed:
Fix ambient factor

SaschaWillems committed:
Clamp specular contribution

SaschaWillems committed:
Fix copyright

SaschaWillems committed:
Enable device extensions required for proper Slang support

SaschaWillems committed:
Enable feature required when using Slang

SaschaWillems commented:
hello_triangle and color_write_enable should also be fixed now.

SaschaWillems committed:
Add CMake option to disable Slang shader compilation

SaschaWillems committed:
Add note on Slang compiler requirements

gary-sweet commented on shaders/conditional_rendering/slang/model.vert.slang:
```diff
@@ -0,0 +1,54 @@
+{
+    VSOutput output;
+    float4 localPos = mul(ubo.view, mul(modelMatrix, float4(input.Pos, 1.0)));
+    output.Normal = input.Normal;
```
The normal calculation doesn't match the glsl version here (I think hlsl is also wrong). I could see lighting differences between glsl and the others, which is why I saw this.

glsl says: `outNormal = mat3(ubo.view * push_constants.model) * inNormal;`

SaschaWillems replied:
You're right. Will fix that :)

SaschaWillems replied:
Fixed for both HLSL and Slang.

gary-sweet reviewed (changes requested):
Looks much better now, but I did still see conditional rendering looking different. This is why.

SaschaWillems committed:
Fix normal calculation

gary-sweet reviewed (dismissed):
Thanks

asuessenbach commented on samples/extensions/color_write_enable/color_write_enable.cpp:
```diff
@@ -245,6 +245,11 @@ void ColorWriteEnable::request_gpu_features(vkb::core::PhysicalDeviceC &gpu)
 		gpu.get_mutable_requested_features().independentBlend = true;
 	}
+
+	if (get_shading_language() == vkb::ShadingLanguage::SLANG)
```
Before requesting this feature, you should check if it is supported. And if that feature is required, you should add some handling for the not supported case.

SaschaWillems replied:
It is required, so there is no other way. I thought the framework is handling this and displays an error message.

P.S. : The line above also does exactly the same and was already present before this PR ;)

asuessenbach replied:
> The line above also does exactly the same and was already present before this PR ;)

The line above does not the same. It checks if `independentBlend` is available, and if so, it requests it.

SaschaWillems replied:
Sorry, missed that. But anyway, there is no fallback if said feature is required, so if the framework gracefully handles this, I'd prefer to not add additional code.

asuessenbach replied:
Sure, the framework handles it gracefully.
But it can just tell you that a feature was not present on device creation. Here you have the opportunity to tell the user which feature is missing. And bail out early.

SaschaWillems replied:
I have added a check to the sample with an appropriate error.

SaschaWillems committed:
Check if required feature is supported and if not exit gracefully

asuessenbach commented on shaders/compute_nbody/slang/particle.frag.slang:
```diff
@@ -0,0 +1,37 @@
+    float2 CenterPos;
+    float PSize : SV_PointSize;
+    // remove
+    float PointSize;
```
Would, instead of `PSize` and `PointSize` something like
`float PointSize : SV_PointSize;`
be enough?

SaschaWillems replied:
Changed to using the point size/coord related built-ins of Slang.

asuessenbach commented on shaders/compute_nbody/slang/particle.vert.slang:
```diff
@@ -0,0 +1,54 @@
+{
+    float4 Pos : SV_POSITION;
+    float GradientPos;
+    float2 CenterPos;
```
Is `CenterPos` some kind of emulation of the `gl_PointCoord`?
I could not find a lot about the semantic name `SV_PointCoord`, but would that be the right approach, somehow?

SaschaWillems replied:
Changed to using the point size/coord related built-ins of Slang.

asuessenbach commented on shaders/compute_nbody/slang/particle.frag.slang:
```diff
@@ -0,0 +1,37 @@
+float4 main(VSOutput input)
+{
+    float3 color = samplerGradientRamp.Sample(float2(input.GradientPos, 0.0)).rgb;
+    float2 PointCoord = (input.Pos.xy - input.CenterPos.xy) / input.PointSize + 0.5;
```
I can't follow the math behind `PointCoord`.
Maybe some comment describing this?

asuessenbach commented on shaders/conditional_rendering/slang/model.vert.slang:
```diff
@@ -0,0 +1,55 @@
+{
+	float3 Pos;
+	float3 Normal;
+	float3 Color;
```
This `Color` is not used. Can it be removed?

asuessenbach commented on shaders/conservative_rasterization/slang/triangle.frag.slang:
```diff
@@ -0,0 +1,28 @@
+[shader("fragment")]
+float4 main(VSOutput input)
+{
+	return float4(input.Color, 1.0);
```
The corresponding glsl-shader also returns a vec4, but just sets rgb.
Shouldn't we fix that right here?

SaschaWillems replied:
I'd prefer to keep this PR focused on Slang. GLSL shaders could be fixed afterwards.

asuessenbach commented on shaders/conservative_rasterization/slang/triangleoverlay.frag.slang:
```diff
@@ -0,0 +1,22 @@
+[shader("fragment")]
+float4 main()
+{
+	return float4(1.0, 1.0, 1.0, 1.0);
```
Here again, the corresponding glsl-shader returns a vec4, but just sets rgb.
Should be fixed there.

SaschaWillems replied:
As noted earlier, we can easily fix GLSL shaders in a different PR.

asuessenbach commented on shaders/debug_utils/slang/bloom.frag.slang:
```diff
@@ -0,0 +1,76 @@
+								0.0118349786570722,
+								0.0073599963704157,
+								0.0043538453346397,
+								0.0024499299678342};
```
If you change the tab width, you see that those numbers are not aligned any more.

SaschaWillems replied:
I'd prefer to keep this PR focused on Slang. GLSL shaders could be fixed afterwards.

asuessenbach commented on shaders/debug_utils/slang/bloom.frag.slang:
```diff
@@ -0,0 +1,76 @@
+
+
+	const float blurScale = 0.003;
+	const float blurStrength = 1.0;
```
These constants are different in the corresponding glsl-shader.

asuessenbach commented on shaders/debug_utils/slang/bloom.frag.slang:
```diff
@@ -0,0 +1,76 @@
+		ar = ts.y / ts.x;
+	}
+
+	float2 P = inUV.yx - float2(0, (25 >> 1) * ar * blurScale);
```
The corresponding glsl-shader uses `weights.length()`, instead of the magic `25`.
Is something like that `length()` not available in slang (and hlsl as well)?

SaschaWillems replied:
Slang `getCount`. Will use that.

asuessenbach replied:
Two lines later, there's also a magic 25...

asuessenbach commented on shaders/debug_utils/slang/bloom.frag.slang:
```diff
@@ -0,0 +1,76 @@
+	{
+        float2 dv = float2(0.0, i * blurScale) * ar;
+        color += samplerColor1.Sample(P + dv) * weights[i] * blurStrength;
+	}
```
Some more tab/space-mixing here.

asuessenbach commented on shaders/debug_utils/slang/composition.frag.slang:
```diff
@@ -0,0 +1,31 @@
+ */
+
+Sampler2D samplerColor0;
+Sampler2D samplerColor1;
```
No `[[vk::binding(1, 0)]]` and `[[vk::binding(2, 0)]]` here?

SaschaWillems replied:
Exactly. It's not required here, as Slang does that implicitly.

asuessenbach commented on shaders/debug_utils/slang/gbuffer.frag.slang:
Mixing tabs and spaces in this file.

asuessenbach commented on shaders/debug_utils/slang/gbuffer.vert.slang:
Also mixing tabs and spaces here.

asuessenbach commented on shaders/debug_utils/slang/gbuffer.vert.slang:
```diff
@@ -0,0 +1,66 @@
+
+	switch(object_type) {
+		case 0: // Skybox
+            worldPos = mul((float4x3) ubo.skyboxModelview, input.Pos).xyz;
```
Why don't you cast `ubo.skyboxModelView` to a `float3x3`?
Or do not cast at all, but extend `input.Pos` to a `float4`?

I don't know, what's more efficient.

asuessenbach commented on shaders/debug_utils/slang/gbuffer.vert.slang:
```diff
@@ -0,0 +1,66 @@
+            output.Pos = mul(ubo.projection, mul(ubo.modelview, float4(localPos, 1.0)));
+			break;
+	}
+	output.Normal = mul((float4x3)ubo.modelview, input.Normal).xyz;
```
Again, a `float4x3` cast.

asuessenbach commented on shaders/debug_utils/slang/gbuffer.vert.slang:
```diff
@@ -0,0 +1,66 @@
+	output.Normal = mul((float4x3)ubo.modelview, input.Normal).xyz;
+	output.UV = input.UV;
+	
+	float3 lightPos = float3(0.0f, -5.0f, 5.0f);
```
The corresponding glsl-shader calculates a different `lightPos`.

asuessenbach commented on shaders/descriptor_buffer_basic/slang/cube.frag.slang:
```diff
@@ -0,0 +1,31 @@
+struct VSOutput
+{
+    float4 Pos : SV_POSITION;
+    float3 Normal;
```
The `Normal` isn't used here.

asuessenbach commented on shaders/dynamic_blending/slang/blending.frag.slang:
```diff
@@ -0,0 +1,50 @@
+{
+    float4 Pos : SV_POSITION;
+    float2 UV;
+    uint ColorOffset;
```
Does the `ColorOffset` need the `nointerpolation` interpolation modifier?

SaschaWillems replied:
That sample wasn't done by me, so no idea. If that's unnecessary/wrong, it needs to be addressed by the original author of said sample.

asuessenbach replied:
Well, I've asked that because you're using that `nointerpolation` in the vertex shader, but not in the fragment shader. I don't think, that's something the original author is supposed to resolve.

In fact, the glsl shader uses the corresponding attribute `flat` the other way around, in the fragment shader but not in the vertex shader.

SaschaWillems replied:
Well, I didn't write that sample. So I'd prefer having the original authors fix those shaders in a later PR. They do work after all.

SaschaWillems replied:
Adjusted

asuessenbach commented on shaders/dynamic_multisample_rasterization/slang/model.frag.slang:
```diff
@@ -0,0 +1,79 @@
+    float4 color = input.BaseColorFactor;
+
+    if (input.BaseTextureIndex != -1)
+        color = Textures[uint(round(input.BaseTextureIndex))].Sample(input.UV);
```
I know, you did not change that part of the code, but why is it needed to round a not interpolated `uint` and cast it to an `uint` again?

SaschaWillems replied:
No idea. I didn't write that sample or the GLSL/HLSL shaders that the Slang variant is based on.

asuessenbach commented on shaders/dynamic_multisample_rasterization/slang/model.frag.slang:
Some tab-and-space mixing in this file.

asuessenbach commented on shaders/dynamic_multisample_rasterization/slang/model.vert.slang:
```diff
@@ -0,0 +1,78 @@
+    float RoughnessFactor;
+    uint BaseTextureIndex;
+    uint NormalTextureIndex;
+    uint MetallicRoughnessTextureIndex;
```
No need to mark the last five members as `nointerpolation`?

SaschaWillems replied:
I just converted that from the GLSL shaders. Those weren't done by me, so needs to be fixed by the sample authors in a separate PR.

asuessenbach replied:
Well, in the glsl fragment shader, those values are marked as `flat`. Which should probably be transcoded to `nointerpolation`?

SaschaWillems replied:
Prob yes. But tbh. I want this PR do be merged as soon as possible and fix things afterwards. Working on a PR for months in my spare time is becoming kinda stressful. Making all the shaders 100% perfect is nigh impossible for me :/

SaschaWillems replied:
Adjusted. Though I don't think it matters, as it seems to suffice if one shader has that modifier. The original GLSL shaders also only have it one of the shaders.

asuessenbach commented on shaders/dynamic_multisample_rasterization/slang/model.vert.slang:
Again, some tab-and-space mixing.

SaschaWillems committed:
Address pull request feedback

Match GLSL blur parameters

SaschaWillems committed:
Address pull request feedback

Ident using spaces

SaschaWillems committed:
Address pull request feedback

Remove unused input

SaschaWillems committed:
Address pull request feedback

Ident using spaces

SaschaWillems committed:
Address pull request feedback

Adjust light position to match GLSL

SaschaWillems committed:
Address pull request feedback

Cast to float3x3

SaschaWillems committed:
Address pull request feedback

Remove unused normal input

SaschaWillems committed:
Address pull request feedback

Get size of blur parameters from array instead of hard-coding

SaschaWillems committed:
Trying to fix clang format

SaschaWillems committed:
Address pull request feedback

Use point size related built-ins

SaschaWillems commented:
I tried to address all of the feedback. I would appreciate if we could merge this on our next call. Minor things can be fixed on separate PRs.

asuessenbach commented on shaders/extended_dynamic_state2/slang/background.vert.slang:
```diff
@@ -0,0 +1,46 @@
+{
+    float4x4 projection;
+    float4x4 view;
+    float tessLevel;
```
The `tessLevel` isn't used here.

SaschaWillems committed:
Address pull request feedback

Use getCount for const array size

asuessenbach commented on shaders/extended_dynamic_state2/slang/baseline.vert.slang:
```diff
@@ -0,0 +1,71 @@
+    output.FragColor   		= color;
+	float4 localPos    		= mul(ubo.view, mul(model, float4(input.Pos, 1.0)));
+    output.Pos 				= mul(ubo.projection, localPos);
+	// @todo
```
What do you want todo here?

asuessenbach commented on shaders/extended_dynamic_state2/slang/baseline.vert.slang:
Some tab/spaces mixing again.

asuessenbach commented on shaders/extended_dynamic_state2/slang/tess.tesc.slang:
```diff
@@ -0,0 +1,79 @@
+
+[shader("hull")]
+[domain("tri")]
+// @todo layout(triangles, equal_spacing, cw) in;
```
What do you have `@todo` here?
And on lie 76 as well?

asuessenbach commented on shaders/extended_dynamic_state2/slang/tess.tese.slang:
```diff
@@ -0,0 +1,63 @@
+
+struct ConstantsHSOutput
+{
+    float TessLevelOuter[4] : SV_TessFactor;
```
in the `tesc.slang`, the `SV_TessFactor` is only 3 floats.

SaschaWillems committed:
Address pull request feedback

SaschaWillems commented:
Thanks for all the feedback. I hope I've it as far as possible for me. Seeing how this PR is growing (both in size and age), I'd like to have this merged in the current state. Minor things can easily be fixed afterwards.

gary-sweet reviewed (dismissed):
I've just spotted some differences in the texture_loading shaders between glsl and both hlsl and slang. Those are clearly pre-existing issues and can easily be fixed after this goes in.

asuessenbach commented on shaders/fragment_shader_barycentric/slang/object.frag.slang:
Some tab/space mixture again.

asuessenbach commented on shaders/fragment_shader_barycentric/slang/object.vert.slang:
Some tab/space mixture again.

asuessenbach commented on shaders/fragment_shader_barycentric/slang/skybox.vert.slang:
```diff
@@ -0,0 +1,43 @@
+VSOutput main(VSInput input)
+{
+    VSOutput output;
+    output.Pos = mul(ubo.projection, float4(mul((float3x3)ubo.modelview, input.Pos.xyz), 1.0));
```
`inputPos` is a `float3`, so no need to select `.xyz` here

asuessenbach commented on shaders/fragment_shading_rate/slang/scene.vert.slang:
```diff
@@ -0,0 +1,66 @@
+    float4x4 projection;
+    float4x4 modelview;
+    float4x4 skybox_modelview;
+	int colorShadingRates;
```
I know, it doesn't matter, but you probably wanted to name this `color_shading_rates`?

asuessenbach commented on shaders/gshader_to_mshader/slang/gshader_to_mshader.frag.slang:
Some tab/space mixing again.

asuessenbach commented on shaders/gshader_to_mshader/slang/gshader_to_mshader.geom.slang:
```diff
@@ -0,0 +1,62 @@
+
+struct GSOutput
+{
+    float4 Pos : SV_POSITION;
```
As you're mixing `SV_Position` and `SV_POSITION` here: I think the "correct" spelling would be `SV_Position`, even though probably all compilers accept `SV_POSITION` as well.

SaschaWillems replied:
Slang is not case-sensitive for these, but yes, I'll try to keep it consistent within shaders.

asuessenbach commented on shaders/gshader_to_mshader/slang/gshader_to_mshader.geom.slang:
```diff
@@ -0,0 +1,62 @@
+void main(triangle VSOutput input[3], inout LineStream<GSOutput> outStream)
+{
+    float normalLength = 0.1f;
+    for (int i = 0; i < 3; i++)
```
The glsl version calculates just one normal line per triangle.

SaschaWillems committed:
Address pull request feedback and cleanup

SaschaWillems committed:
Address pull request feedback

SaschaWillems commented:
I have addressed all open feedback. For those that reviewed it, please (re)approve so we can merge on the next call.

SaschaWillems committed:
Clean up existing shaders

Remove attributes not required by Slang

asuessenbach commented on shaders/gshader_to_mshader/slang/gshader_to_mshader.geom.slang:
The `maxvertexcount` now is 2.

asuessenbach commented on shaders/hdr/hlsl/gbuffer.frag.hlsl:
```diff
@@ -97,7 +97,8 @@ FSOutput main(VSOutput input)
                 fresnel += F0;
 
-                float spec = (fresnel * geoAtt) / (NdotV * NdotL * 3.14);
+				// Note: clamp to zero to mitigate any divide by zero
```
This clamping is questionable.
But as it's also done in the glsl shader, it could be addressed by the original author in a separate PR.

gary-sweet replied:
It might be questionable, but without it our hardware produces the wrong answer.

SaschaWillems replied:
Based on gary's comment I'll keep this in code.

asuessenbach commented on shaders/hdr/slang/bloom.frag.slang:
Some tab/space mixing again.

asuessenbach commented on shaders/hdr/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,128 @@
+{
+    float3 Pos;
+    float3 Normal;
+};
```
The `VSInput` isn't used in this fragment shader.

asuessenbach commented on shaders/hdr/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,128 @@
+		{
+			float3 wViewVec = mul((float3x3)uboMatrices.inverseModelView, normalize(input.ViewVec)).xyz;
+			float3 normal = normalize(input.Normal);
+			float3 wNormal = mul((float3x3)uboMatrices.inverseModelView, normal).xyz;
```
The results of these operations are `float3` anyways. No need to select `.xyz`.

asuessenbach commented on shaders/hdr/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,128 @@
+		case 2: // Refract
+		{
+			float3 wViewVec = mul((float4x3)uboMatrices.inverseModelView, normalize(input.ViewVec)).xyz;
+            float3 wNormal = mul((float4x3)uboMatrices.inverseModelView, input.Normal).xyz;
```
When you'd cast to a `float3x3`, you would not need to select `.xyz`.

asuessenbach commented on shaders/hdr/slang/gbuffer.frag.slang:
Some tab/spaces mixing again.

asuessenbach commented on shaders/hdr/slang/gbuffer.vert.slang:
```diff
@@ -0,0 +1,65 @@
+			output.Pos = mul(ubo.projection, mul(ubo.modelview, float4(input.Pos.xyz * ubo.modelscale, 1.0)));
+			break;
+    }
+    output.Normal = mul((float4x3)ubo.modelview, input.Normal).xyz;
```
If you cast to `float3x3` (3 times here), you don't need to select `.xyz`.

asuessenbach commented on shaders/instancing/slang/instancing.frag.slang:
Some tab/spaces mixing.

asuessenbach commented on shaders/instancing/slang/instancing.vert.slang:
```diff
@@ -0,0 +1,103 @@
+    float4 pos = float4((locPos.xyz * input.instanceScale) + input.instancePos, 1.0);
+
+    output.Pos = mul(ubo.projection, mul(ubo.modelview, mul(gRotMat, pos)));
+    output.Normal = mul((float3x3)mul(ubo.modelview, gRotMat), mul(rotMat, input.Normal));
```
The `rotMat` is the inverse (here: tranpose) of that in the glsl shader. But that's correct, as it's used to left-multiply, while it's right-multiplied in the glsl shader.
But `gRotMat` is also the inverse of that in the glsl shader, but they are used identically in both shaders. That is, one of them is probably wrong.

SaschaWillems replied:
This would require a fix on the sample side. Visually they do match (or only differ very little), so IMO it's fine. The HLSL shader has the same issue.

asuessenbach reviewed (commented):
Some more comments...

asuessenbach commented on shaders/logic_op_dynamic_state/slang/background.vert.slang:
```diff
@@ -0,0 +1,44 @@
+{
+    VSOutput output;
+	output.UVW = input.Pos;
+    output.Pos = mul(ubo.projection, mul(ubo.view, float4(input.Pos.xyz * 10.0, 1.0)));
```
`input.Pos` is a `float3`, so you don't have to select `.xyz`.

asuessenbach commented on shaders/logic_op_dynamic_state/slang/baseline.vert.slang:
Some tab/space mixing.

SaschaWillems commented:
@asuessenbach: Please let me know when you're finished with your review. It's easier for me to go through the remaining issues one last time.

asuessenbach commented on shaders/patch_control_points/slang/tess.tese.slang:
```diff
@@ -0,0 +1,62 @@
+	float3 Color;
+};
+
+/// layout(triangles, equal_spacing, cw) in;
```
This comment can probably be removed.

asuessenbach commented on shaders/ray_queries/slang/ray_shadow.frag.slang:
```diff
@@ -0,0 +1,132 @@
+*/
+bool intersects_light(float3 light_origin, float3 pos)
+{
+	const float tmin = 0.01, tmax = 1.0;
```
Both, slang and hlsl, use `tmax = 1.0`. glsl uses `tmax = 1000`.
I assume, the glsl value is wrong, or at least unnecessary large?

SaschaWillems replied:
Yeah, the GLSL value is unnecessary large. But it works, so IMO the difference is fine and can be fixed in a separate PR.

asuessenbach commented on shaders/ray_tracing_extended/slang/closesthit.rchit.slang:
```diff
@@ -0,0 +1,148 @@
+}
+
+[shader("closesthit")]
+void main(inout Payload hitValue, in Attributes Attribs)
```
Maybe make the `Attribs` lower case.

asuessenbach commented on shaders/ray_tracing_extended/slang/raygen.rgen.slang:
Some tab/spaces mixing.

asuessenbach commented on shaders/ray_tracing_position_fetch/slang/raygen.rgen.slang:
```diff
@@ -0,0 +1,54 @@
+	rayDesc.TMin = 0.001;
+	rayDesc.TMax = 10000.0;
+
+    Payload payload;
```
Don't you need to initialize the payload?

asuessenbach commented on shaders/separate_image_sampler/slang/separate_image_sampler.vert.slang:
```diff
@@ -0,0 +1,48 @@
+    float4x4 model;
+    float4 viewPos;
+    float locSpeed;
+    float globSpeed;
```
Where are those `locSpeed` and `globSpeed` are coming from?

asuessenbach commented on shaders/shader_debugprintf/slang/scene.frag.slang:
Some tab/space mixing.

asuessenbach commented on shaders/shader_debugprintf/slang/scene.vert.slang:
Some tab/space mixing.

asuessenbach commented on shaders/shader_debugprintf/slang/scene.vert.slang:
```diff
@@ -0,0 +1,73 @@
+
+	switch(object_type) {
+		case 0: // Skysphere                     
+            tmpPos = mul((float3x3) ubo.skybox_modelview, input.Pos.xyz);
```
`input.Pos` is a `float3`. No need to select `.xyz`.

asuessenbach commented on shaders/sparse_image/slang/sparse.vert.slang:
```diff
@@ -0,0 +1,45 @@
+
+struct VSInput
+{
+    float3 Pos : POSITION0;
```
The C++-source just feeds two float2 (r32g32) as vertex input.
Does that somehow work nonetheless?

asuessenbach commented on shaders/sparse_image/slang/sparse.vert.slang:
```diff
@@ -0,0 +1,45 @@
+{
+    VSOutput output;
+    output.UV = input.UV;
+    output.Pos = mul(mvp.proj, mul(mvp.view, mul(mvp.model, float4(input.Pos.xyz, 1.0))));
```
`input.Pos` is a `float3` -> no need to select `.xyz`.

asuessenbach commented on shaders/synchronization_2/slang/particle.frag.slang:
The same comments regarding `PointSize` and `CenterPos` as for the compute_nbody sample.

asuessenbach commented on shaders/synchronization_2/slang/particle_calculate.comp.slang:
```diff
@@ -0,0 +1,87 @@
+};
+ConstantBuffer<UBO> ubo;
+
+[[vk::constant_id(1)]] const int SHARED_DATA_SIZE = 512;
```
The glsl shader uses 1024 as the `SHARED_DATA_SIZE`. hsls and slang use 512.
Should probably be unified?

asuessenbach commented on shaders/terrain_tessellation/slang/terrain.frag.slang:
```diff
@@ -0,0 +1,79 @@
+	layers[2] = float2(45.0, 80.0);
+	layers[3] = float2(75.0, 100.0);
+	layers[4] = float2(95.0, 140.0);
+	layers[5] = float2(140.0, 190.0);
```
The glsl shader uses different values for layers[4] and layers[5]. Should probably be unified.

asuessenbach commented on shaders/texture_loading/slang/texture.frag.slang:
```diff
@@ -0,0 +1,41 @@
+ * limitations under the License.
+ */
+
+[[vk::binding(1, 0)]] Sampler2D SamplerColor;
```
`SamplerColor` -> `samplerColor` ?

asuessenbach commented on shaders/texture_loading/slang/texture.frag.slang:
```diff
@@ -0,0 +1,41 @@
+    float3 V = normalize(input.ViewVec);
+    float3 R = reflect(-L, N);
+    float3 diffuse = max(dot(N, L), 0.0).xxx;
+    float3 specular = (pow(max(dot(R, V), 0.0), 16.0)).xxx * color.a;
```
Is this `.xxx` needed?
Does slang not support multiplication of a scalar with a vector?

asuessenbach commented on shaders/texture_loading/slang/texture.vert.slang:
```diff
@@ -0,0 +1,57 @@
+VSOutput main(VSInput input)
+{
+    VSOutput output;
+    float3 vecPos = mul((float3x3) ubo.model, input.Pos.xyz).xyz;
```
No need to `.xyz` here. The result of the multiplication should already be a `float3`.

asuessenbach commented on shaders/texture_loading/slang/texture.vert.slang:
```diff
@@ -0,0 +1,57 @@
+    output.UV = input.UV;
+    output.LodBias = ubo.lodBias;
+    output.Pos = mul(ubo.projection, mul(ubo.model, float4(input.Pos.xyz, 1.0)));
+    output.Normal = mul((float3x3) ubo.model, input.Normal);
```
The glsl shader takes the `inverse(transpose(ubo.model))`
As long as the 3x3-part of the `ubo.model` is a pure rotation, you can skip that. But if it contained some shearing or scaling, you can not.

SaschaWillems replied:
It does not contain shearing or scaling afaik. HLSL/Slang do not have inverse, hence why I moved it to the CPU side for those samples that require it. This one doesn't.

asuessenbach commented on shaders/texture_mipmap_generation/slang/texture.vert.slang:
```diff
@@ -0,0 +1,46 @@
+struct UBO
+{
+    float4x4 projection;
+    float4x4 model;
```
I think, `modelview`, instead of `model`, like it's named in the glsl shader, would be a better name.

asuessenbach commented on shaders/timeline_semaphore/slang/game_of_life_mutate.comp.slang:
```diff
@@ -0,0 +1,40 @@
+[numthreads(8, 8, 1)]
+void main(uint3 globalInvocationID: SV_DispatchThreadID, uniform float counter)
+{
+    float4 v = ImageInput.Load(int3(globalInvocationID.xy, 0));
```
`Load` on a `Sampler2D` takes an `int3` as input?

SaschaWillems replied:
Yes, just like it does in HLSL. See https://learn.microsoft.com/de-de/windows/win32/direct3dhlsl/dx-graphics-hlsl-to-load

asuessenbach commented on shaders/timeline_semaphore/slang/game_of_life_update.comp.slang:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+
+    ImageOutput[int2(globalInvocationID.xy)] = float4(total, 0.0);
```
Do you need both, the `int2`-cast and the `.xy` selection?

asuessenbach commented on shaders/vertex_dynamic_state/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,113 @@
+
+        case 1: // Reflect
+			{
+                float3 wViewVec = mul((float3x3) uboMatrices.inverseModelView, normalize(input.ViewVec)).xyz;
```
No need to select `.xyz`.

asuessenbach commented on shaders/vertex_dynamic_state/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,113 @@
+			{
+                float3 wViewVec = mul((float3x3) uboMatrices.inverseModelView, normalize(input.ViewVec)).xyz;
+                float3 normal = normalize(input.Normal);
+                float3 wNormal = mul((float3x3) uboMatrices.inverseModelView, normal).xyz;
```
As above.

asuessenbach commented on shaders/vertex_dynamic_state/slang/gbuffer.frag.slang:
```diff
@@ -0,0 +1,113 @@
+        case 2: // Refract
+			{
+                float3 wViewVec = mul((float4x3) uboMatrices.inverseModelView, normalize(input.ViewVec)).xyz;
+                float3 wNormal = mul((float4x3)uboMatrices.inverseModelView, input.Normal).xyz;
```
Why do you cast to a `float4x3` and then select `.xyz`?
You could cast to a `float3x3` (as above), and don't explicitly select `.xyz`.

asuessenbach commented on shaders/vertex_dynamic_state/slang/gbuffer.vert.slang:
```diff
@@ -0,0 +1,69 @@
+    switch (type)
+    {
+        case 0: // Skybox
+            worldPos = mul((float4x3) ubo.skyboxModelview, input.Pos).xyz;
```
Again, the combination of `float4x3` and `.xyz` selection.

asuessenbach commented on shaders/vertex_dynamic_state/slang/gbuffer.vert.slang:
```diff
@@ -0,0 +1,69 @@
+            output.Pos = mul(ubo.projection, mul(ubo.modelview, float4(input.Pos.xyz * ubo.modelscale, 1.0)));
+            break;
+    }
+    output.Normal = mul((float4x3) ubo.modelview, input.Normal).xyz;
```
As above.

asuessenbach reviewed (commented):
Alright now, those are my final comments on this PR!

SaschaWillems commented:
@asuessenbach : Just a quick note that this PR is getting to  point where github has issues loading it. Let me know when you're done with your review and I'll try to address all of feedback, but after that we should move remaining things to a separate issue. Looks like the comments are mostly cosmetic (and partially also wrong in GLSL/HLSL). I'd like to finish this PR at some point...

SaschaWillems committed:
Spaces instead of tabs

SaschaWillems committed:
Address PR feedback
